rts/idris_rts.h

```diff
@@ -25,68 +23,115 @@
 
 // Closures
 typedef enum {
-    CT_CON, CT_ARRAY, CT_INT, CT_BIGINT, CT_FLOAT, CT_STRING, CT_STROFFSET,
-    CT_BITS8, CT_BITS16, CT_BITS32, CT_BITS64, CT_UNIT, CT_PTR, CT_REF,
-    CT_FWD, CT_MANAGEDPTR, CT_RAWDATA, CT_CDATA
+    CT_CON, CT_ARRAY, CT_INT, CT_BIGINT,
+    CT_FLOAT, CT_STRING, CT_STROFFSET, CT_BITS8,
+    CT_BITS16, CT_BITS32, CT_BITS64, CT_PTR,
+    CT_REF, CT_FWD, CT_MANAGEDPTR, CT_RAWDATA,
+    CT_CDATA
 } ClosureType;
 
-typedef struct Closure *VAL;
-
-// A constructor, consisting of a tag, an arity (16 bits each of the
-// tag_arity field) and arguments
-typedef struct {
-    uint32_t tag_arity;
-    VAL args[];
-} con;
-
-// An array; similar to a constructor but with a length, and contents
-// initialised to NULL (high level Idris programs are responsible for
-// initialising them properly)
-typedef struct {
-    uint32_t length;
-    VAL content[];
-} array;
-
-typedef struct {
-    VAL str;
-    size_t offset;
-} StrOffset;
-
-typedef struct {
-    char* str;
-    size_t len; // Cached strlen (we do 'strlen' a lot)
+typedef struct Hdr {
+    uint8_t ty;
+    uint8_t u8;
+    uint16_t u16;
+    uint32_t sz;
+} Hdr;
+
+typedef struct Val {
+    Hdr hdr;
+} Val;
+
+typedef struct Val * VAL;
+
+typedef struct Con {
+  Hdr hdr;
+  uint32_t tag;
+  uint32_t arity;
+  VAL args[0];
+} Con;
+
+typedef struct Array {
+  Hdr hdr;
+  VAL array[0];
+} Array;
+
+typedef struct Int {
+  Hdr hdr;
+  int i;
+} Int;
+
+typedef struct BigInt {
+  Hdr hdr;
+  char big[0];
+} BigInt;
+
+typedef struct Float {
+  Hdr hdr;
+  double f;
+} Float;
+
+typedef struct String {
+  Hdr hdr;
+  size_t slen;
+#define _null hdr.u8
```

Comment:
Remove this and substitute `hdr.u8` in the use sites and put in a comment that String uses the hdr.u8 field for null.